File: LeetCode/0934-shortest-bridge/0934-shortest-bridge.py
class Solution:
    def dfs(self, grid: List[List[int]], i: int, j: int, q: deque):
        n = len(grid)
        if i < 0 or j < 0 or i >= n or j >= n or grid[i][j] != 1:
            return
        grid[i][j] = 2
        q.append((i, j))
        self.dfs(grid, i + 1, j, q)
        self.dfs(grid, i - 1, j, q)
        self.dfs(grid, i, j + 1, q)
        self.dfs(grid, i, j - 1, q)

    def shortestBridge(self, grid: List[List[int]]) -> int:
        n = len(grid)
        q = deque()
        found = False
        
        for i in range(n):
            if found:
                break
            for j in range(n):
                if grid[i][j] == 1:
                    self.dfs(grid, i, j, q)
                    found = True
                    break
        
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        steps = 0
        
        while q:
            size = len(q)
            for _ in range(size):
                x, y = q.popleft()
                for dx, dy in dirs:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < n and 0 <= ny < n:
                        if grid[nx][ny] == 1:
                            return steps
                        if grid[nx][ny] == 0:
                            grid[nx][ny] = 2
                            q.append((nx, ny))
            steps += 1
        
        return -1

File: LeetCode/0934-shortest-bridge/0934-shortest-bridge.java
class Solution {
    public void dfs(int[][] grid, int i, int j, Queue<int[]> q) {
        int n = grid.length;
        if (i < 0 || j < 0 || i >= n || j >= n || grid[i][j] != 1) {
            return;
        }
        grid[i][j] = 2;
        q.offer(new int[]{i, j});
        dfs(grid, i + 1, j, q);
        dfs(grid, i - 1, j, q);
        dfs(grid, i, j + 1, q);
        dfs(grid, i, j - 1, q);
    }

    public int shortestBridge(int[][] grid) {
        int n = grid.length;
        Queue<int[]> q = new LinkedList<>();
        boolean found = false;
        
        for (int i = 0; i < n; i++) {
            if (found) {
                break;
            }
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    dfs(grid, i, j, q);
                    found = true;
                    break;
                }
            }
        }
        
        int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int steps = 0;
        
        while (!q.isEmpty()) {
            int size = q.size();
            for (int k = 0; k < size; k++) {
                int[] xy = q.poll();
                for (int[] dir : dirs) {
                    int nx = xy[0] + dir[0];
                    int ny = xy[1] + dir[1];
                    if (nx >= 0 && nx < n && ny >= 0 && ny < n) {
                        if (grid[nx][ny] == 1) {
                            return steps;
                        }
                        if (grid[nx][ny] == 0) {
                            grid[nx][ny] = 2;
                            q.offer(new int[]{nx, ny});
                        }
                    }
                }
            }
            steps++;
        }
        return -1;
    }
}